src/append/journald/field.rs

```diff
@@ -0,0 +1,195 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This field is derived from https://github.com/swsnr/systemd-journal-logger.rs/blob/v2.2.0/src/fields.rs.
+
+//! Write well-formated journal fields to buffers.
+
+use std::fmt::Arguments;
+use std::io::Write;
+
+use log::kv::Value;
+
+pub(super) enum FieldName<'a> {
+    WellFormed(&'a str),
+    WriteEscaped(&'a str),
+}
+
+/// Whether `c` is a valid character in the key of a journal field.
+///
+/// Journal field keys may only contain ASCII uppercase letters A to Z,
+/// numbers 0 to 9 and the underscore.
+fn is_valid_key_char(c: char) -> bool {
+    matches!(c, 'A'..='Z' | '0'..='9' | '_')
+}
+
+/// Write an escaped `key` for use in a systemd journal field.
+///
+/// See [`super::Journald`] for the rules.
+fn write_escaped_key(key: &str, buffer: &mut Vec<u8>) {
+    // Key length is limited to 64 bytes
+    let mut remaining = 64;
+
+    let escaped = key
+        .to_ascii_uppercase()
+        .replace(|c| !is_valid_key_char(c), "_");
+
+    if escaped.starts_with(|c: char| matches!(c, '_' | '0'..='9')) {
+        buffer.extend_from_slice(b"ESCAPED_");
+        remaining -= 8;
+    }
+
+    for b in escaped.into_bytes() {
+        if remaining == 0 {
+            break;
+        }
+        buffer.push(b);
+        remaining -= 1;
+    }
+}
+
+fn put_field_name(buffer: &mut Vec<u8>, name: FieldName<'_>) {
+    match name {
+        FieldName::WellFormed(name) => buffer.extend_from_slice(name.as_bytes()),
+        FieldName::WriteEscaped("") => buffer.extend_from_slice(b"EMPTY"),
+        FieldName::WriteEscaped(name) => write_escaped_key(name, buffer),
+    }
+}
+
+pub(super) trait PutAsFieldValue {
+    fn put_field_value(self, buffer: &mut Vec<u8>);
+}
+
+impl PutAsFieldValue for &[u8] {
+    fn put_field_value(self, buffer: &mut Vec<u8>) {
+        buffer.extend_from_slice(self)
+    }
+}
+
+impl PutAsFieldValue for &Arguments<'_> {
+    fn put_field_value(self, buffer: &mut Vec<u8>) {
+        match self.as_str() {
+            Some(s) => buffer.extend_from_slice(s.as_bytes()),
+            None => {
+                // SAFETY: no more than an allocate-less version
+                //  buffer.extend_from_slice(format!("{}", self))
+                write!(buffer, "{}", self).unwrap()
+            }
+        }
+    }
+}
+
+impl PutAsFieldValue for Value<'_> {
+    fn put_field_value(self, buffer: &mut Vec<u8>) {
+        // SAFETY: no more than an allocate-less version
+        //  buffer.extend_from_slice(format!("{}", self))
+        write!(buffer, "{}", self).unwrap();
+    }
+}
```

Comment:
@swsnr 

And for this leaving TODO "We can probably write the value more efficiently by visiting it?", I noticed that we actually format integers as their string representation instead of bytes. Thus visiting the value only add extra code without certain efficiency improvement.